doc/developer/translating-enums.md

```diff
@@ -0,0 +1,278 @@
+--   of the generated SAWCore ENUMNAME__case function to appropriate arguments.
+importCase ::
+  (HasCallStack) =>
+  SharedContext -> Env ->
+  C.Type -> C.Expr -> Map C.Ident C.CaseAlt -> Maybe C.CaseAlt -> IO Term
+```
+
+## Design Decisions
+
+Here we capture the design principles and design decisions that the implementation follows.
+  * Enums are represented by generic sums (i.e., nested Eithers) in
+    SAWCore.
+    - It was originally felt that it could result in faster
+      development time than the approach of mapping to a SAWCore
+      datatype definition.  In hindsight, it was merely trading
+      understanding one part of the codebase with another.  We are not
+      sure if we gained or lost time with this decision.
+    - Mapping to SAWCore datatype defitions would be a future
+      improvement, as it gives us injective types (we won't have the
+      possibility of two distinct enum types mapping to the same
+      SAWCore representation).
+    
+  * We prefer 1. over 2. over 3:
+    1. Using code in the SAWCore prelude.
+    2. Generating (type-correct, type-checked) code that is added to
+       the SAWCore environment when we see enum declarations.
+    3. Generating code when we see case expressions.
+
+  * When possible we want to be generating common SAWCore abstractions
+    and generating code that is generic over the SAWCore `ListSort`
+    structure: a (type level) list of types.
+
+## Useful Definitions from the SAWCore Prelude
+
+We will be making heavy use of `ListSort` in the SAWCore prelude:
+
+```
+-- A list of types, i.e. `List (sort 0)` if `List` was universe polymorphic
+data ListSort : sort 1
+  where {
+    LS_Nil : ListSort;
+    LS_Cons : sort 0 -> ListSort -> ListSort;
+  }
+  
+```
+
+For which we also make use of
+
+```
+listSortGet : ListSort -> Nat -> sort 0;
+listSortDrop : ListSort -> Nat -> ListSort;
+```
+
+We also use this generic N-ary Sum, `EithersV`
+
+```
+-- EithersV (a simpler version of Eithers): A right-nested sequence of
+-- Either types, defined as
+--   EithersV [] = Void
+--   EithersV (tp:tps) = Either tp (EithersV tps)
+EithersV : ListSort -> sort 0;
+EithersV = ListSort__rec
+            (\ (_:ListSort) -> sort 0)
+            Void
+            (\ (tp:sort 0) (_:ListSort) (rec:sort 0) ->
+               (Either tp rec));
+```
+
+The eliminator for `EithersV` (the case for this N-Sum) is
+
+```
+-- An eliminator for the EithersV type
+eithersV : (a:sort 0) -> (es:FunsTo a) -> EithersV (FunsToIns a es) -> a;
+```
+
+We will also need this from the prelude:
+
+```
+-- Form the multiple arrow type a1 -> ... -> an -> b
+arrowsType : ListSort -> sort 0 -> sort 0;
+arrowsType as b =
+  ListSort__rec (\ (_:ListSort) -> sort 0) b
+    (\ (a:sort 0) (_:ListSort) (rec:sort 0) -> a -> rec)
+    as;
+```    
+    
+In the below, we'll be relying also on `TupleR` (not in the prelude)
+for defining a right-nested product:
```

Comment:
I'm a bit confused by this. If `TupleR` is not in the prelude, how are you making use of it in the translation?

Is the idea that you are using `TupleR` in the high-level explanation below, but in the actual implementation, you are inlining the definition of `TupleR`? If so, it would be helpful to say as much.